### ISSUE TITLE
How to use custom likelihood (VAE like ) 

### ISSUE THREAD

**Issue by nicola144** [cornellius-gp/gpytorch#1344]
I was wondering how to achieve defining a likelihood where the output is given by some trainable neural net (eg. , a Gaussian with mean and covariance given by a neural net) . 

More specifically, for my problem I am trying to extend the  SVDKL example : 
https://docs.gpytorch.ai/en/v1.2.1/examples/06_PyTorch_NN_Integration_DKL/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.html

where I want to output an image - so , adding a neural net after the GP layer, in a VAE-like style, and train everything end to end.  If this is already possible to do , I would appreaciate a suggestion. Thanks

**Comment by jacobrgardner** [comment]
So if you mean you want to pass the output of a layer of GPs through a NN, when you call `rsample` on the (probably batch) `MultivariateNormal` output of a GP, you'll get gradients through the GP to whatever parameters were involved in getting you that output. In that regard, you can kind of do anything you want with the GP output as long as whatever you do is also differentiable.

Do you have some more specific formulation for what you are trying to do?

**Comment by nicola144** [comment]
Thanks a lot for your reply. 
I see what you mean ; I am not sure that is what I would like tho.  I was referring to : I want the output of the GP to be an image like tensor , and I wanted to achieve that by writing a "gpytorch.likelihoods" object, and train everything end to end.  For example, this likelihood object could be a gpytorch "MultitaskGaussianLikelihood" , but where mean and covariance are parameterized by a trainable net. 
 I could do what you said , but I believe it would be different - i.e., over a training step , I don't just want to sample from a GP and transform it -  I want to define a GP posterior with a complicated likelihood as said above.  Just to clarify ,  if I did what you said,  does GPytorch already perform the "reparameterization trick" automatically ? 
Thanks a lot

**Comment by gpleiss** [comment]
A code example would be helpful here.

**Comment by nicola144** [comment]
It turns out what I meant to achieve was a model defined in this paper from you guys : https://arxiv.org/pdf/1905.13697.pdf  

At least, I meant specifically the model in 3.1.5 but with fixed "M" .   (I would also combine it with DKL but that's not important for what I need to ask  )  

Are some of the models in the paper integrated into GPytorch ?  If not (since the paper does not have code) , could you give some directions as to where I would need to go to create such a likelihood ?  
Thanks

**Comment by nicola144** [comment]
It turns out exactly what I meant and need is done in this paper from you guys : https://arxiv.org/pdf/1905.13697v1.pdf 

Could you give some directions as to where I would need to subclass/extend to implement this neural likelihood ? 
This doesn't seem to be into GPytorch at the moment from what I can see, and the paper does not have code. 
Thanks

**Comment by gpleiss** [comment]
You should create your own custom likelihood. The `forward` method accepts samples from a (multi-output) multivariate normal and computes the likelihood of those samples. See [the softmax likelihood](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/likelihoods/softmax_likelihood.py) for an example.

In your case, you'll want the samples in the forward method to be passed through a neural network before returning a NormalDistribution object.

Also see [the likelihood documentation](https://docs.gpytorch.ai/en/stable/likelihoods.html#gpytorch.likelihoods.Likelihood.forward).
